README
                                 Duh DRAW

                   'Making a DOS-Free world a reality'
     
      DuhDraw, written by Ben Fowler and maintained by Pekka Enberg, is a
      Linux based Open Source clone of TheDRAW, a DOS based ASCII art editor.

  Authors

      Copyright (c) 1996 Ben Fowler
      Copyright (c) 2002-2003 Pekka Enberg

  License

      Duh Draw is free software. You can distribute and/or modify it under the
      terms of the GNU General Public License version 2 as published by the
      Free Software Foundation. Please refer to the file LICENSE for details.

  Using Duh Draw

      Maximum Columns - 80 (Fixed)
      Maximum Lines - 1000
 
      Home - Beginning of Line/ANSI.
      End  - End of Line/ANSI.
      Page Up - Up 23 lines.
      Page Down - Down 23 lines.

      Delete - Delete character under cursor.

      Insert - Toggle insert mode.
      When insert mode is active 'In' appears in the status line.

      F1 thru F10 - character in currently selected high ascii set.
      Shift F1 thru Shift F10 - Select highascii set 1 thru 10.

      High ascii sets 11, 12, 13, 14, 15 are accessed by pressing Alt->
      or Alt-<.

      Alt-A toggles attribute drawing mode.
      When in attribute drawing mode 'Attr' will appear in the status
      line. In this mode moving the cursor causes the character
      newly under the cursor to acquire the current foreground
      and background colors.
 
      Alt-B does the block mode thing ALA XxxXxxx.
      Block mode does not highlight the selected block area yet. It will.
      You can now copy blocks.
      Block mode mirror and invert work correctly and
      mirror (or invert) highascii characters.
      Translate will change all instances of one color to another
      while preserving bright/normal intensity.
      Selecting fill mode will bring up a sub menu to fill
      using the current foreground color,background color,both foreground
      and background color or any character you type.
      You may use one of the highascii characters in the selected set
      or press Alt-C to use  the last character picked up with Alt-P.

      Alt-C draws using the last character picked up with Alt-P in the
      current foreground/background color.

      Alt-D toggles line drawing mode.
      Line drawing uses whatever high ascii set you have selected.
      Using high ascii sets above set 4 produces odd results.
      Line drawing mode is not up to snuff yet but will be.
    
      Alt-E prompts if you wish to erase the current page.
      (Currently there is only one page).
   
      Alt-F brings up the color bar where you select fore and back
      colors. Cursor keys to select. Enter to accept.
      
      Alt-H  Help screen.
      Alt-I  Insert Line.
      Alt-J  Insert Column.
      Alt-K  Kill (Delete) Column.  
      Alt-L  Load file.
      Alt-P  Pickup character under cursor.
      Alt-S  Save file.
      Alt-U  Use color under cursor.
      Alt-X  Exit Duh DRAW.
      Alt-Y  Yank (Delete) line.

  Limitations

      Animation mode is not supported.


HISTORY
This file represents the history of Duh DRAW, it's drawn from all sorts
of historical files and lore.


02-Feb-1996  [From the Whatdone file]

 Woah. Had a big week. 
 'ANSI Master' got a new name (DuhDraw) Courtesy HalfLife :) Thanks Mate.
 All work on DuhDraw came to a screeching halt due the fact that
 top was showing it at 96.5% cpu usage. :)
 The problem was traced back to the fact that the function getchar()
 was continously returning -1 to indicate no keypress available.
 After some probing into the nature of tty modes,and just as I was about
 to do a major kludge of a signal driven keyboard driver,I discovered
 cbreak() in the curses library.
 Also thanks to Halflife,there is now the likely hood that pulldown menu's
 will exist in a future version (using LibMenu).
 An MDI (Multiple Document Interface) version of DuhDraw is in the
 works takng much better advantage of LibCurses,something I hardly
 understood at the time DuhDraw was begun.

 In the meantime,block save has been added. Block load is still in the
 works,as is block copy,which is still not working yet.
 It's probable that block move/copy will be implemented by copying a section
 to a 'clipboard' and then pasting clipboard in etc.

 Looking into the nature of gpm (mouse driver) to add much needed 
 mouse support. Surely this will become a major point of kludging
 betwixt the bsd and linux versions.

 Which reminds me,thanks to kmem for working on the BSD port of DuhDraw.

 Added new opening logo screen courtesy Deathblow of The Dirty Dozen.

 Animation mode is still a sticky issue. 
 The problem is not the scanning modes (gate,pyramid,circle etc..)
 but the process of appending color changes,cursor pos'es after the fact.
 We'll see how I feel (and thus how it is finnaly implemented).

 DuhDraw is still a somewhat stupid program in that it assumes VT100+
 term capabilites and 25 line by 80 char display. Running it with a dumb 
 terminal will produce truely ugly results and probably result in a 
 few panicked moments of futile attempts to exit the program. :)

 'all we do is sit around and play with little colored blocks' - Cavalier


06-May-1996  [Reconstructed History]
v2.6.96      This is the latest date on the .lsm file.  This appears to
             actually be the date it was submitted to sunsite.unc.edu.
             The code's origin remain a mystery, other than the name
             Ben Fowler and an obsolete email address.

04-Apr-2002  After repeated attempts to contact "the" Ben Fowler, I managed
v2.6.96      to accidently reach "a" Ben Fowler who says he's been fielding
             responses for the last few years about DuhDraw and that there
             is much confusion, not to mention no one's heard from Ben. 

             The project has been on FreshMeat.net for quite some time and
             appears to be abandonware.  As such, I've opted to pick up
             the task of maintaining it and applying contributed patches.

             Walt Stoneburner, the new maintainer, can be reached at
             <wls@wwco.com> and the home page for the project is
             currently http://www.wwco.com/~wls/opensource/duhdraw.php

             Ben, should you return, I'll gladly hand the torch back.

04-Apr-2002  This version compiles under gcc 3.0.1 or better using
v2.7.0       ncurses.  Known bugs: when the terminal exits, you may
             not have echoing.  For now, do this:
               $ stty sane ; clear ; reset

             Fixes include: 
             - Making the C compiler optimizations more agressive
             - Using NCURSES instead of CURSES
             - Include proper headers
             - General code cleanup and type conversion issues
             - Fix switch statement problems
             - Prevent yes/no from falling out without a return value
             - I/O Flushing so screen reflects state
             - Resolved nested comments
             - Fix bulk initializtion
             - Removed unused variables
             - Remove compiler warnings

05-Apr-2002  Thanks to Pressly Dowler <squirrel@galaxy7.dhs.org> for
v2.7.1       contributing a patch which resolved restoring the terminal
             state to normal when the program ends.  I applied this to
             both DuhDraw and Ansi.

             Thanks to Tamas TEVESZ <ice@extreme.hu> for contributing
             a patch to DuhDraw which allowed it to compile with GCC;
             I had made a goof and use G++ instead.

             New features:
               In ANSI, you can now use space to do a page down and
               return to go down one line.

31-Aug-2002  Pekka Enberg <Pekka.Enberg@cs.helsinki.fi> - patch
v2.7.2       to fix crash when trying to read a file that wasn't open.

03-Sep-2002  Pekka convinced me that the code would be better maintained
v2.7.3       if it were indented to a better standard.  Consequently, I've
             run all the sources through indent using the GNU style.

             Formatting changes only.

04-Sep-2002  Pekka Enberg <Pekka.Enberg@cs.helsinki.fi> - patch
v2.7.4       to remove copy-block code, which always crashed.

04-Sep-2002  Pekka Enberg <Pekka.Enberg@cs.helsinki.fi> reimplements
v2.7.5       copy-block code from scratch.

             Note, there still appears to be a problem where under
             some obscure situation you can get a segmentation
             fault.  However, for the most part this works much
             better than before.

             I've included a new TODO file which addresses
             coding and security issues I'd like to get taken
             care of.

             Minor cleanup was also done to the code in regards
             to get it to compile without warnings via -Wall.

05-Sep-2002  Pekka Enberg <Pekka.Enberg@cs.helsinki.fi> removes
2.7.6        dead commented code in an effort to make the code
             base more maintainable.

05-Sep-2002  Pekka Enberg <Pekka.Enberg@cs.helsinki.fi> repairs

ddlogo.h
/*
 * c data file converted from ansi file by ansitoc (c) 1996 Ben Fowler
 */

char screendata[]={ 27,91,63,55,104,27,91,52,48,109,27,91,50,
                    74,27,91,50,59,49,72,27,91,51,59,49,72,
                    27,91,52,59,49,72,27,91,53,59,49,72,27,
                    91,54,59,49,72,27,91,54,67,27,91,48,59,
                    49,59,51,48,109,220,219,219,219,223,223,32,32,
                    223,32,27,91,115,13,10,27,91,117,32,32,223,
                    27,91,49,51,67,223,27,91,49,52,67,223,27,
                    91,49,51,67,223,32,32,32,32,223,32,32,223,
                    223,219,219,219,220,27,91,55,59,49,72,32,32,
                    32,32,220,219,223,223,27,91,54,53,67,223,223,
                    219,220,27,91,56,59,49,72,32,27,91,115,13,
                    10,27,91,117,32,32,32,219,223,27,91,48,59,
                    51,53,109,46,27,91,49,59,51,48,109,253,27,
                    91,48,59,51,53,109,46,27,91,49,109,223,223,
                    223,223,223,219,27,91,52,53,109,178,27,91,55,
                    67,27,91,52,48,109,219,27,91,115,13,10,27,
                    91,117,27,91,52,53,109,178,27,91,55,67,27,
                    91,52,48,109,219,27,91,52,53,109,178,27,91,
                    52,67,27,91,52,48,109,223,223,223,223,223,219,
                    27,91,52,53,109,178,27,91,50,67,27,91,52,
                    48,109,223,223,223,223,223,219,27,91,115,13,10,
                    27,91,117,27,91,52,53,109,178,27,91,50,67,
                    27,91,52,48,109,223,223,223,223,223,219,27,91,
                    52,53,109,178,27,91,55,67,27,91,52,48,109,
                    219,27,91,52,53,109,178,27,91,49,67,27,91,
                    51,55,59,52,48,109,250,32,27,91,51,48,109,
                    223,219,13,10,27,91,57,59,49,72,32,32,32,
                    32,178,32,27,91,51,53,109,96,27,91,51,55,
                    109,33,27,91,51,53,109,39,27,91,51,49,59,
                    52,53,109,178,177,178,27,91,50,67,27,91,51,
                    53,59,52,48,109,223,27,91,52,53,109,176,27,
                    91,52,48,109,27,91,115,13,10,27,91,117,27,
                    91,50,67,27,91,51,49,59,52,53,109,178,177,
                    178,27,91,50,67,27,91,51,53,59,52,48,109,
                    223,27,91,52,53,109,176,27,91,50,67,27,91,
                    51,49,109,178,177,178,27,91,50,67,27,91,51,
                    53,59,52,48,109,223,27,91,115,13,10,27,91,
                    117,27,91,52,53,109,176,27,91,52,67,27,91,
                    51,49,109,178,177,178,27,91,50,67,27,91,51,
                    53,59,52,48,109,223,27,91,52,53,109,176,27,
                    91,50,67,27,91,51,49,109,178,177,178,27,91,
                    50,67,27,91,52,48,109,27,91,115,13,10,27,
                    91,117,27,91,51,53,59,52,48,109,223,27,91,
                    52,53,109,176,27,91,50,67,27,91,51,49,109,
                    178,177,178,27,91,50,67,27,91,51,53,59,52,
                    48,109,223,27,91,52,53,109,176,27,91,50,67,
                    27,91,51,49,109,178,177,178,27,91,52,48,109,
                    27,91,115,13,10,27,91,117,27,91,50,67,27,
                    91,51,53,109,223,27,91,52,53,109,176,27,91,
                    49,67,27,91,51,55,59,52,48,109,46,32,32,
                    27,91,51,48,109,178,27,91,49,48,59,49,72,
                    32,32,32,32,177,32,32,27,91,51,55,109,46,
                    32,27,91,115,13,10,27,91,117,27,91,51,49,
                    59,52,53,109,177,176,177,27,91,51,67,27,91,
                    51,48,109,176,27,91,50,67,27,91,51,49,109,
                    177,176,177,27,91,51,67,27,91,51,48,109,176,
                    27,91,50,67,27,91,51,49,109,177,176,177,27,
                    91,52,48,109,27,91,115,13,10,27,91,117,27,
                    91,51,53,59,52,48,109,220,220,220,27,91,51,
                    48,59,52,53,109,176,27,91,52,67,27,91,51,
                    49,109,177,176,177,27,91,51,67,27,91,51,48,
                    109,176,27,91,50,67,27,91,51,49,109,177,176,
                    177,27,91,52,48,109,27,91,115,13,10,27,91,
                    117,27,91,51,53,59,52,48,109,220,220,27,91,
                    52,53,109,223,27,91,48,59,51,53,109,223,32,
                    32,27,91,49,59,51,49,59,52,53,109,177,176,
                    177,27,91,51,53,59,52,48,109,220,220,220,27,
                    91,51,48,59,52,53,109,176,27,91,52,48,109,
                    27,91,115,13,10,27,91,117,27,91,50,67,27,
                    91,51,49,59,52,53,109,177,176,177,27,91,49,
                    67,27,91,51,53,59,52,48,109,220,32,27,91,
                    51,48,59,52,53,109,176,27,91,52,67,27,91,
                    52,48,109,177,27,91,49,49,59,49,72,32,32,
                    32,32,177,32,32,32,32,27,91,115,13,10,27,
                    91,117,27,91,52,53,109,176,178,176,27,91,51,
                    67,177,27,91,50,67,176,178,176,27,91,51,67,
                    177,27,91,50,67,176,178,176,27,91,51,67,177,
                    27,91,52,67,176,178,176,27,91,51,67,177,27,
                    91,50,67,176,178,176,27,91,52,48,109,27,91,
                    115,13,10,27,91,117,27,91,50,67,223,27,91,
                    52,53,109,177,27,91,50,67,176,178,176,27,91,
                    51,67,177,27,91,50,67,176,177,176,27,91,49,
                    67,177,27,91,49,67,177,27,91,51,53,59,52,
                    48,109,46,27,91,51,55,109,33,27,91,51,53,
                    109,46,32,27,91,115,13,10,27,91,117,27,91,
                    51,48,109,177,27,91,49,50,59,49,72,27,91,
                    55,67,27,91,51,55,109,250,32,27,91,51,48,
                    109,220,220,220,220,220,220,27,91,52,53,109,178,
                    27,91,50,67,27,91,52,48,109,220,220,220,220,
                    220,220,27,91,115,13,10,27,91,117,27,91,52,
                    53,109,178,27,91,50,67,27,91,52,48,109,220,
                    220,220,32,32,32,27,91,52,53,109,178,27,91,
                    52,67,27,91,52,48,109,220,220,220,220,220,220,
                    27,91,52,53,109,178,27,91,50,67,27,91,52,
                    48,109,220,220,220,32,32,32,27,91,115,13,10,
                    27,91,117,27,91,52,53,109,178,27,91,50,67,
                    27,91,52,48,109,220,220,220,32,32,32,27,91,
                    52,53,109,178,27,91,50,67,27,91,52,48,109,
                    220,220,220,220,223,220,27,91,52,53,109,178,27,
                    91,48,59,51,53,109,96,27,91,49,59,51,48,
                    109,168,27,91,115,13,10,27,91,117,27,91,48,
                    59,51,53,109,39,27,91,49,51,59,49,72,32,
                    32,32,32,27,91,49,59,51,48,109,176,32,32,
                    219,223,27,91,53,54,67,100,98,91,116,100,100,
                    93,223,219,32,32,176,27,91,49,52,59,49,72,
                    27,91,55,67,223,221,27,91,115,13,10,27,91,
                    117,27,91,52,53,109,32,32,32,32,32,32,32,
                    32,32,32,32,32,32,27,91,51,53,109,68,85,
                    72,68,82,65,87,32,70,79,82,32,76,105,78,
                    85,88,32,27,91,51,48,109,40,27,91,51,53,
                    109,99,27,91,51,48,109,41,27,91,52,48,109,
                    27,91,115,13,10,27,91,117,27,91,52,53,109,
                    32,27,91,51,53,109,66,69,78,32,70,79,87,
                    76,69,82,32,49,57,57,54,32,32,32,32,32,
                    32,32,32,32,32,32,32,32,27,91,51,48,59,
                    52,48,109,222,223,27,91,49,53,59,49,72,32,
                    32,32,32,176,32,32,32,27,91,115,13,10,27,
                    91,117,219,220,220,220,32,220,220,32,220,32,32,
                    32,220,27,91,55,67,220,27,91,49,49,67,220,
                    27,91,55,67,220,27,91,56,67,220,27,91,53,
                    67,220,32,220,220,32,220,220,220,220,219,32,32,
                    32,176,27,91,48,109,13,10,0 };


ansitoc.c
/****************************************************************************
 *
 *          ansitoc.c -  convert an ANSI file to c data format.
 *
 *                    (c) Copyright 1996 Ben Fowler
 *                          All Rights Reserved
 *
 ****************************************************************************/

#include <stdio.h>
#include <string.h>

/* convert ansi to c data statement. ansi would be displayed with
 *  printf("%s",screendata);
 */

int
main (int argc, char *argv[])
{
    int count, c;
    FILE *in;
    FILE *out;
    char infile[80];
    char outfile[80];

    if (argc != 3)
        {
            printf ("\n%s (c) Copyright 1996 Ben Fowler.\n", argv[0]);
            printf ("Purpose : Convert an ANSI file to c data statements.\n");
            printf ("Usage : %s infile outfile\n\n", argv[0]);
            return 1;
        }

    strcpy (infile, argv[1]);
    strcpy (outfile, argv[2]);
    in = fopen (infile, "rb");
    out = fopen (outfile, "wb");

    count = 0;

    fprintf (out, "/* \n");
    fprintf (out,
             "* c data file converted from ansi file by ansitoc (c) 1996 Ben Fowler\n");
    fprintf (out, "*/\n\n");
    fprintf (out, "char screendata[]={ ");

    while ((c = fgetc (in)) != EOF)
        {
            fprintf (out, "%d,", c);
            count++;
            if (count > 12)
                {
                    count = 0;
                    fprintf (out, "\n                   ");
                }
        }

    fprintf (out, "0 };\n");
    fclose (in);
    fclose (out);

    return 0;

}


ansi.c
/*****************************************************************************
 *                          ANSI Viewer for Linux
 *                               Ben Fowler
 *                              December  1995
 *     Derived largely in part from /usr/src/linux/driver/char/console.c
 *
 *
 ****************************************************************************/

#include <stdio.h>
#include <string.h>
#include <curses.h>
#include <stdlib.h>
#include "ansi-esc.h"

static int rows = 25;			/* default terminal lines */
static int cols = 80;			/* default terminal columns */

static int x = 0;
static int y = 0;
static int pos = 0;
static int par[8];
static int splos = 0;

#define EDIT_BUFFER_HEIGHT 1000

static unsigned int * editbuffer;

static int ques = 0;
static int npar = 0;
static int attr = 7;
static int maxread = 0;
static int ishome = 0;
static int lattr;
static int lastline;

static void
clearbuf (unsigned int * buffer, unsigned long size)
{
    int i;

    for (i = 0; i < size; i++)
        {
            buffer[i] = 0x0720;	/* default color ' ' */
        }
}

static void
printattr (void)
{
    int fg;
    int bg;
    int bold;
    int flash;

    flash = 25;
    fg = (attr & 0x07);
    bg = (attr >> 4) & 0x07;
    fg = fg + 30;
    bg = bg + 40;
    bold = 0;
    if (attr & 0x08)
        bold = 1;
    if (attr & 0x80)
        flash = 5;
    printf ("\e[0;%d;%d;%d;%dm", bold, flash, fg, bg);
}

static void
print_line (void)			/* print one line to screen */
{
    int co;
    for (co = 0; co < cols; co++)
        {
            attr = (editbuffer[pos + co] >> 8);
            if (lattr != attr)
                {
                    printattr ();
                    lattr = attr;
                }
            putc (editbuffer[pos + co] & 0xFF, stdout);
        }
}

static void
ansiscrollup (void)
{
    if ((pos - (rows * cols)) >= 0)
        {
            splos = pos;
            pos = pos - (rows * cols);
            printf ("\e[0;0H\e[L\e[0;0H");	/* insert a line */
            print_line ();
            pos = splos - cols;
            ishome = 0;
        }
    else
        ishome = 1;
}

static void
ansiscrolldown (void)
{
    if ((pos + cols) >= maxread)
        return;
    if (lastline < rows)
        return;

    {
        printf ("\e[%d;0H\n", rows);
        pos += cols;
        print_line ();
        ishome = 0;
    }
}

static void
home (void)
{
    int i, zz;
    if (ishome == 1)		/* if already at top skippit */
        return;
    zz = rows;
    if (lastline < rows)
        zz = lastline;
    printf ("\e[2J\e[0;0H");
    for (i = 0; i < zz; i++)
        {
            pos = i * cols;
            print_line ();
        }
    ishome = 1;
}

static void
end (void)
{
    int i, zz;
    if (lastline < rows)
        return;
    if ((pos + cols) >= maxread)	/* if already at end skippit */
        return;

    zz = lastline - rows;
    if (zz < 0)
        zz = 0;
    zz = zz * cols;
    printf ("\e[2J\e[0;0H");
    for (i = 0; i < rows; i++)
        {
            pos = zz + (i * cols);
            print_line ();
        }
    ishome = 0;
}

static void
pagedown (void)
{
    int i;
    i = pos / cols;
    if ((lastline - i) <= rows)
        {
            end ();
        }
    else
        {
            printf ("\e[2J\e[0;0H");
            for (i = 0; i < rows; i++)
                {
                    pos = pos + cols;
                    print_line ();
                }
            ishome = 0;
        }
}

static void
pageup (void)
{
    int i;
    if ((pos - (2 * rows * cols)) >= 0)
        {
            printf ("\e[2J\e[0;0H");
            pos = pos - (2 * rows * cols);
            for (i = 0; i < rows; i++)
                {
                    pos = pos + cols;
                    print_line ();		/* leave it pointing at last line */
                }
        }
    else
        home ();
}

static void
scrollit (void)
{
    int c, ok;
    int ex = 0;
    int scrollstate = ESnormal;

    printf ("\e[?25l");		/* turn cursor off */
    printf ("\e[2J\e[0;0H");	/* clear screen and home cursor */
    printf ("\e(U");		/* set linux virtual console to IBM_PC graphics set */
    ishome = 0;

    lastline = pos / cols;
    c = lastline * cols;
    ok = pos - c;
    if (ok > 0)
        lastline++;
    maxread = pos;
    pos = 0;
    c = 0;
    ok = 0;
    home ();

    while (fflush (stdout) || (((ex != 1) && (c = getchar ()) != 255)))
        {
            switch (c)
                {
                case 7:
                    continue;
                case 8:
                    continue;
                case 9:		/* tab key */
                    continue;
                case 11:
                case 12:
                    continue;
                case 10:
                case 13:
                    ansiscrolldown ();
                    continue;
                case 14:
                    continue;
                case 15:
                    continue;
                case 24:
                case 26:
                    scrollstate = ESnormal;
                    continue;
                case 27:
                    scrollstate = ESesc;
                    continue;
                case 4:
                case 'Q':
                case 'q':
                case 'X':
                case 'x':
                    ex = 1;
                    continue;
                case ' ':
                    pagedown ();
                    continue;
                case 127:
                    continue;
                case 128 + 27:
                    scrollstate = ESsquare;
                    continue;
                }

            switch (scrollstate)
                {
                case ESesc:
                    scrollstate = ESnormal;
                    switch (c)
                        {
                        case '[':
                            scrollstate = ESsquare;
                            continue;
                        case ']':
                            scrollstate = ESnonstd;
                            continue;
                        case '%':
                            scrollstate = ESpercent;
                            continue;
                        case 'E':
                            continue;
                        case 'M':
                            continue;
                        case 'D':
                            continue;
                        case 'H':
                            continue;
                        case 'Z':
                            continue;
                        case '7':
                            continue;
                        case '8':
                            continue;
                        case '(':
                            scrollstate = ESsetG0;
                            continue;
                        case ')':
                            scrollstate = ESsetG1;
                            continue;
                        case '#':
                            scrollstate = EShash;
                            continue;
                        case 'c':
                            continue;
                        case '>':		/* Numeric keypad */
                            continue;
                        case '=':		/* Appl. keypad */
                            continue;
                        }
                    continue;
                case ESnonstd:
                    if (c == 'P')
                        {			/* palette escape sequence */
                            for (npar = 0; npar < 7; npar++)
                                par[npar] = 0;
                            npar = 0;
                            scrollstate = ESpalette;
                            continue;
                        }
                    else if (c == 'R')
                        {			/* reset palette */
                            scrollstate = ESnormal;
                        }
                    else
                        scrollstate = ESnormal;
                    continue;
                case ESpalette:
                    if ((c >= '0' && c <= '9') || (c >= 'A' && c <= 'F')
                        || (c >= 'a' && c <= 'f'))
                        {
                            par[npar++] = (c > '9' ? (c & 0xDF) - 'A' + 10 : c - '0');
                        }
                    else
                        scrollstate = ESnormal;
                    continue;
                case ESsquare:
                    for (npar = 0; npar < 7; npar++)
                        par[npar] = 0;
                    npar = 0;
                    scrollstate = ESgetpars;
                    if (c == '[')
                        {			/* Function key */
                            scrollstate = ESfunckey;
                            continue;
                        }
                    ques = (c == '?');
                    if (ques)
                        continue;
                case ESgetpars:
                    if (c == ';' && npar < 6)
                        {
                            npar++;
                            continue;
                        }
                    else if (c >= '0' && c <= '9')
                        {
                            par[npar] *= 10;
                            par[npar] += c - '0';
                            continue;
                        }
                    else
                        scrollstate = ESgotpars;
                case ESgotpars:
                    scrollstate = ESnormal;
                    switch (c)
                        {
                        case 'h':
                            continue;
                        case 'l':
                            continue;
                        case 'n':
                            continue;
                        }
                    if (ques)
                        {
                            ques = 0;
                            continue;
                        }
                    switch (c)
                        {
                        case 'G':
                        case '`':
                            continue;
                        case 'A':
                            ansiscrollup ();
                            continue;
                        case 'B':
                        case 'e':
                            ansiscrolldown ();
                            continue;
                        case 'C':
                        case 'a':
                            continue;
                        case 'D':
                            continue;
                        case 'E':
                            continue;
                        case 'F':
                            continue;
                        case 'd':
                            continue;
                        case 'H':
                        case 'f':
                            continue;
                        case 'J':
                            continue;
                        case 'K':
                            continue;
                        case 'L':
                            continue;
                        case 'M':
                            continue;
                        case 'P':
                            continue;
                        case 'c':
                            continue;
                        case 'g':
                            continue;
                        case 'm':
                            continue;
                        case 'q':		/* DECLL - but only 3 leds */
                            continue;
                        case 'r':
                            continue;
                        case 's':
                            continue;
                        case 'u':
                            continue;
                        case 'X':
                            continue;
                        case '@':
                            continue;
                        case ']':		/* setterm functions */
                            continue;
                        case '~':		/* home keys */
                            if (par[0] == 1)
                                home ();
                            if (par[0] == 4)
                                end ();
                            if (par[0] == 5)
                                pageup ();
                            if (par[0] == 6)
                                pagedown ();
                            continue;
                        }
                    continue;
                case ESpercent:
                    scrollstate = ESnormal;
                    continue;
                case ESfunckey:
                    scrollstate = ESnormal;
                    continue;
                case EShash:
                    scrollstate = ESnormal;
                    continue;
                case ESsetG0:
                    scrollstate = ESnormal;
                    continue;
                case ESsetG1:
                    scrollstate = ESnormal;
                    continue;
                default:
                    scrollstate = ESnormal;
                }
        }				/* end switch(scrollstate) */

    printf ("\e(B");		/* set unix character set */
    printf ("\e[2J\e[0;0H\n");
    printf
        ("\e[0;1;32;40m Linux Ansi Viewer by \e[0;1;37;40mBen Fowler\e[0m\n\n");
    printf ("\e[?25h");		/* turn cursor on */
}				/* end scrollit() */

static void
usage (void)
{
    printf ("Usage: ansi [FILE]...\n");
    printf ("View ANSI escape coded FILEs on your terminal.\n\n");
    printf ("Commands while viewing:\n");
    printf ("  Up    - Scroll up\n");
    printf ("  Down  - Scroll down\n");
    printf ("  Pg Up - Up one page\n");
    printf ("  Pg Dn - Down one page\n");
    printf ("  Home  - Go to top of ansi\n");
    printf ("  End   - Go to end of ansi\n");
    printf ("  X     - Exit viewer / View next file\n");
}

void
view_ansi (FILE * fp)
{
    savetty ();
    initscr ();
    nonl ();
    cbreak ();
    noecho ();
    clear ();

    getmaxyx(stdscr, rows, cols);

    x = 0;
    y = 0;
    pos = 0;
    splos = 0;

    editbuffer = (unsigned int *) malloc(sizeof(unsigned int) * cols * EDIT_BUFFER_HEIGHT);
    if (NULL == editbuffer)
        {
            fprintf (stderr, "Could not allocate memory for buffer.\n");
            exit (1);
        }

    clearbuf (editbuffer, cols * EDIT_BUFFER_HEIGHT);
    pos = ansi_esc_translate (fp, editbuffer, cols, EDIT_BUFFER_HEIGHT);
    scrollit ();
    free (editbuffer);

    echo ();
    nl ();
    resetty ();
    endwin ();
}

int
main (int argc, char *argv[])
{
    FILE *fp;
    int arg;

    if (argc < 2)
        {
            usage ();
            return EXIT_SUCCESS;
        }

    for (arg = 1; arg < argc; arg++)
        {
            if ((fp = fopen (argv[arg], "r")) != NULL)
                {
                    view_ansi (fp);
                    fclose (fp);
                }
            else
                {
                    printf ("\nCould not open '%s'.\n", argv[arg]);
                    usage ();
                    return EXIT_FAILURE;
                }
        }

    return EXIT_SUCCESS;
}
